routing/index_graph_starter.hpp

```diff
@@ -21,136 +26,180 @@ class IndexGraphStarter final
+  // If segment is real returns true and does not modify segment.
+  // If segment is part of real converts it to real and returns true.
+  // Otherwise returns false and does not modify segment.
+  bool ConvertToReal(Segment & segment) const;
+  Junction const & GetJunction(Segment const & segment, bool front) const;
+  Junction const & GetRouteJunction(std::vector<Segment> const & route, size_t pointIndex) const;
+  m2::PointD const & GetPoint(Segment const & segment, bool front) const;
+  size_t GetNumFakeSegments() const { return m_fake.m_segmentToVertex.size(); }
+
+  std::set<NumMwmId> GetMwms() const;
+
+  void GetEdgesList(Segment const & segment, bool isOutgoing,
+                    std::vector<SegmentEdge> & edges) const;
+
+  void GetOutgoingEdgesList(TVertexType const & segment, std::vector<TEdgeType> & edges) const
   {
     GetEdgesList(segment, true /* isOutgoing */, edges);
   }
 
-  void GetIngoingEdgesList(TVertexType const & segment, vector<TEdgeType> & edges)
+  void GetIngoingEdgesList(TVertexType const & segment, std::vector<TEdgeType> & edges) const
   {
     GetEdgesList(segment, false /* isOutgoing */, edges);
   }
 
-  RouteWeight HeuristicCostEstimate(TVertexType const & from, TVertexType const & to)
+  RouteWeight HeuristicCostEstimate(TVertexType const & from, TVertexType const & to) const
   {
     return RouteWeight(m_graph.GetEstimator().CalcHeuristic(GetPoint(from, true /* front */),
                                                             GetPoint(to, true /* front */)));
   }
 
-  double CalcSegmentWeight(Segment const & segment) const
-  {
-    return m_graph.GetEstimator().CalcSegmentWeight(
-        segment, m_graph.GetRoadGeometry(segment.GetMwmId(), segment.GetFeatureId()));
-  }
+  double CalcSegmentWeight(Segment const & segment) const;
+  double CalcRouteSegmentWeight(std::vector<Segment> const & route, size_t segmentIndex) const;
 
-  bool IsLeap(NumMwmId mwmId) const
-  {
-    return mwmId != kFakeNumMwmId && mwmId != m_start.GetMwmId() && mwmId != m_finish.GetMwmId() &&
-           m_graph.GetEstimator().LeapIsAllowed(mwmId);
-  }
+  bool IsLeap(NumMwmId mwmId) const;
 
   static bool IsFakeSegment(Segment const & segment)
   {
     return segment.GetFeatureId() == kFakeFeatureId;
   }
 
 private:
-  void GetFakeToNormalEdges(FakeVertex const & fakeVertex, bool isOutgoing,
-                            vector<SegmentEdge> & edges);
-  void GetFakeToNormalEdge(FakeVertex const & fakeVertex, bool forward,
-                           vector<SegmentEdge> & edges);
-  void GetNormalToFakeEdge(Segment const & segment, FakeVertex const & fakeVertex,
-                           Segment const & fakeSegment, bool isOutgoing,
-                           vector<SegmentEdge> & edges);
+  static Segment GetFakeSegment(uint32_t i)
+  {
+    return Segment(kFakeNumMwmId, kFakeFeatureId, i, false);
+  }
+
+  class FakeVertex final
+  {
+  public:
+    enum class Type
+    {
+      PureFake,
```

Comment:
Do you need to distinguish between PureFake and Projection for now? If not, please, join these types.

Reply:
done